File: android/app/src/main/kotlin/com/stillmoment/domain/models/Introduction.kt
package com.stillmoment.domain.models

import java.util.Locale

/**
 * Represents an optional introduction audio that plays at the beginning of a meditation session.
 *
 * Introductions are bundled audio files (e.g., guided breathing exercises) that play
 * after the start gong and before the silent meditation phase. They are language-specific
 * and can be configured in the timer settings.
 *
 * @property id Unique, language-independent identifier (e.g., "breath")
 * @property nameEnglish English display name
 * @property nameGerman German display name
 * @property durationSeconds Duration of the introduction audio in seconds
 * @property availableLanguages Languages for which audio files are available
 * @property filenamePattern Pattern with {lang} placeholder for language-specific audio files
 */
data class Introduction(
    val id: String,
    val nameEnglish: String,
    val nameGerman: String,
    val durationSeconds: Int,
    val availableLanguages: List<String>,
    val filenamePattern: String
) {
    /**
     * Returns the localized name based on current device locale.
     */
    val localizedName: String
        get() = when (currentLanguage) {
            "de" -> nameGerman
            else -> nameEnglish
        }

    /**
     * Returns the audio resource filename for the given language.
     *
     * @param language Language code (e.g., "de")
     * @return Resource filename or null if language not available
     */
    fun audioFilename(language: String): String? {
        if (!availableLanguages.contains(language)) return null
        return filenamePattern.replace("{lang}", language)
    }

    /**
     * Formatted duration string (e.g., "1:35").
     */
    val formattedDuration: String
        get() {
            val minutes = durationSeconds / 60
            val seconds = durationSeconds % 60
            return String.format(Locale.ROOT, "%d:%02d", minutes, seconds)
        }

    companion object {
        /** Override for testing — set to a language code to bypass Locale.getDefault(). */
        var languageOverride: String? = null

        /** Returns the current device language code (or test override). */
        val currentLanguage: String
            get() = languageOverride ?: Locale.getDefault().language

        /** All registered introductions. */
        val allIntroductions: List<Introduction> = listOf(
            Introduction(
                id = "breath",
                nameEnglish = "Breathing Exercise",
                nameGerman = "Atemübung",
                durationSeconds = 95,
                availableLanguages = listOf("de"),
                filenamePattern = "intro_breath_{lang}"
            )
        )

        /**
         * Returns introductions available for the current device language.
         */
        fun availableForCurrentLanguage(): List<Introduction> {
            val lang = currentLanguage
            return allIntroductions.filter { it.availableLanguages.contains(lang) }
        }

        /**
         * Checks if any introductions are available for the current device language.
         */
        val hasAvailableIntroductions: Boolean
            get() = availableForCurrentLanguage().isNotEmpty()

        /**
         * Finds an introduction by ID.
         *
         * @param id The introduction ID to search for
         * @return The matching Introduction or null
         */
        fun find(id: String): Introduction? = allIntroductions.find { it.id == id }

        /**
         * Checks if an introduction is available for the current device language.
         *
         * @param id The introduction ID to check
         * @return true if the introduction exists and has audio for the current language
         */
        fun isAvailableForCurrentLanguage(id: String): Boolean {
            val intro = find(id) ?: return false
            return intro.availableLanguages.contains(currentLanguage)
        }

        /**
         * Returns the audio resource filename for an introduction in the current device language.
         *
         * @param id The introduction ID
         * @return Resource filename or null if not available
         */
        fun audioFilenameForCurrentLanguage(id: String): String? {
            val intro = find(id) ?: return null
            return intro.audioFilename(currentLanguage)
        }
    }
}

File: ios/StillMoment/Domain/Models/Introduction.swift
//
//  Introduction.swift
//  Still Moment
//
//  Domain Model - Introduction for Meditation Timer
//

import Foundation

/// Represents an optional introduction audio that plays at the beginning of a meditation session
///
/// Introductions are bundled audio files (e.g., guided breathing exercises) that play
/// after the start gong and before the silent meditation phase. They are language-specific
/// and can be configured in the timer settings.
///
/// Example usage:
/// ```swift
/// let intro = Introduction.breath
/// let filename = intro.audioFilename(for: "de") // "intro-breath-de.mp3"
/// ```
struct Introduction: Identifiable, Equatable {
    /// Unique, language-independent identifier (e.g., "breath")
    let id: String

    /// Localized display name (resolved via NSLocalizedString at creation time)
    let name: String

    /// Duration of the introduction audio in seconds, per language
    let durationByLanguage: [String: Int]

    /// Languages for which audio files are available
    let availableLanguages: [String]

    /// Filename pattern with `{lang}` placeholder (e.g., "intro-breath-{lang}.mp3")
    let filenamePattern: String

    /// Returns the audio filename for a specific language
    /// - Parameter language: Language code (e.g., "de", "en")
    /// - Returns: Resolved filename or nil if language not available
    func audioFilename(for language: String) -> String? {
        guard self.availableLanguages.contains(language) else {
            return nil
        }
        return self.filenamePattern.replacingOccurrences(of: "{lang}", with: language)
    }

    /// Maximum duration across all languages (used for minimum timer calculation)
    var durationSeconds: Int {
        self.durationByLanguage.values.max() ?? 0
    }

    /// Duration for a specific language, or max if language not found
    func durationSeconds(for language: String) -> Int {
        self.durationByLanguage[language] ?? self.durationSeconds
    }

    /// Formatted duration string (e.g., "1:35") — uses max duration across languages
    var formattedDuration: String {
        let minutes = self.durationSeconds / 60
        let seconds = self.durationSeconds % 60
        return String(format: "%d:%02d", minutes, seconds)
    }
}

// MARK: - Static Introduction Definitions

extension Introduction {
    /// Breathing exercise introduction (first available introduction)
    static let breath = Introduction(
        id: "breath",
        name: NSLocalizedString("introduction.breath.name", comment: ""),
        durationByLanguage: ["de": 95, "en": 93],
        availableLanguages: ["de", "en"],
        filenamePattern: "intro-breath-{lang}.mp3"
    )

    /// All registered introductions
    static let allIntroductions: [Introduction] = [
        breath
    ]
}

// MARK: - Registry Methods

extension Introduction {
    /// Override for testing — set to a language code to bypass `Locale.current`.
    /// Must be reset to `nil` in test tearDown.
    static var languageOverride: String?

    /// Returns the current device language code (or test override)
    static var currentLanguage: String {
        self.languageOverride ?? Locale.current.language.languageCode?.identifier ?? "en"
    }

    /// Returns introductions available for the current device language
    static func availableForCurrentLanguage() -> [Introduction] {
        let lang = self.currentLanguage
        return self.allIntroductions.filter { $0.availableLanguages.contains(lang) }
    }

    /// Checks if any introductions are available for the current device language
    static var hasAvailableIntroductions: Bool {
        !self.availableForCurrentLanguage().isEmpty
    }

    /// Finds an introduction by ID
    /// - Parameter id: The introduction ID to search for
    /// - Returns: The matching Introduction or nil
    static func find(byId id: String) -> Introduction? {
        self.allIntroductions.first { $0.id == id }
    }

    /// Checks if an introduction is available for the current device language
    /// - Parameter id: The introduction ID to check
    /// - Returns: true if the introduction exists and has audio for the current language
    static func isAvailableForCurrentLanguage(_ id: String) -> Bool {
        guard let intro = find(byId: id) else {
            return false
        }
        return intro.availableLanguages.contains(self.currentLanguage)
    }

    /// Returns the audio filename for an introduction in the current device language
    /// - Parameter id: The introduction ID
    /// - Returns: Resolved filename or nil if not available
    static func audioFilenameForCurrentLanguage(_ id: String) -> String? {
        guard let intro = find(byId: id) else {
            return nil
        }
        return intro.audioFilename(for: self.currentLanguage)
    }
}
